docs/alternativesemantics.md

```diff
@@ -0,0 +1,202 @@
+C1,r1,iso ~>_r C2,_,mut = C2,r1,mut
+C1,r1,imm ~>_r C2,_,mut = C2,r1,imm
+C1,r1,mut ~>_r C2,_,iso = C2,r,mut
+C1,r1,iso ~>_r C2,_,iso = C2,r,mut
+C1,r1,imm ~>_r C2,_,iso = C2,r1,imm
+C1,r1,mut ~>_r C2,_,imm = C2,r1,imm
+C1,r1,iso ~>_r C2,_,imm = C2,r1,imm
+C1,r1,imm ~>_r C2,_,imm = C2,r1,imm
+```
+**[Could be a relational definition to allow for unrestricted r on the right hand side.]**
+
+Extracting viewpoint adaptation
+```
+C1,r1,mut <~_r C2,_,mut = C2,r1,mut
+C1,r1,iso <~_r C2,_,mut = C2,r1,mut
+C1,r1,imm <~_r C2,_,mut = C2,r1,imm
+C1,r1,mut <~_r C2,_,iso = C2,r,iso
+C1,r1,iso <~_r C2,_,iso = C2,r,iso
+C1,r1,imm <~_r C2,_,iso = C2,r1,imm
+C1,r1,mut <~_r C2,_,imm = C2,r1,imm
+C1,r1,iso <~_r C2,_,imm = C2,r1,imm
+C1,r1,imm <~_r C2,_,imm = C2,r1,imm
+```
+
+The most important difference between extracting and aliasing
+viewpoint adaption is how they behave on `iso` fields. Extracting provides an `iso` in the supplied region, and aliasing provides a `mut` in the supplied region.
+
+The definitions take a region parameter so that "freshness" can be suitably defined. 
+
+### Type Rules
+
+Aliasing assignment generates a fresh region
+if it follows an `iso`, if it follows an `imm` then it is in the immutable region, otherwise it is in the same region. 
+
+Read versus Write types for a field
+```
+   field_r(tau1 | tau2, f) = field_r(tau1, f) | field_r(tau2, f)
+   field_r(tau1 & tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+   field_w(tau1 | tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+   field_w(tau1 & tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+```
+]
+
+```
+field_R(f, tau_x) = tau_f
+fresh r
+-------------------------------------
+G |- z = x.f -| G, z: tau_x ->r tau_f
+```
+
+Extraction uses a fresh region
+```
+G(x) : tau_x
+G(y) : tau_y
+field_w(f, tau_x) = tau_f_w
+field_r(f, tau_x) = tau_f_r
+exists r'. tau_y <: tau_x <~r' tau_f_w
+fresh r
+--------------------------------------
+G |- z = (x.f = y) -| G, z: tau_x <~r tau_f_r
+```
+Is y still accessible?  If it was an iso, then no.  But otherwise, yes.  So remove y, if we can prove it was not an iso?
+
+
+Method call?
+
+No longer has to invalidate any regions, may consume locals of type `iso`.
+```
+  foo (x: iso&C) {...}
+
+  var y = new C
+  foo (y)
+  foo (y) /// Error `y` was invalidate by previous line.
+```
+
+
+
+##  Work required
+
+* Runtime support for delaying reclaimation
+* Runtime support for delaying message send
+* Feasability work for optimisations 
+    - What can we optimise?
+    - What user annotation might we provide?
+    - 
+* Type system implementation 
+* Optimisation
+
+## Concerns
+
+### Additional root sets
+
+Not tracking regions and subregions precisely means we may need to be more careful with codegen for the additional root set.  For instance, we can use have the same region multiple times, which means we use the region additional roots in a globally stack based way.  I.e.
+```
+  r1.add_root(a)
+  r2.add_root(b)
+  ...
+  r1.remove_root(a)
+  ...

```

Comment:
I'm suspecting that this area is where the problem lies. It'd be easier to see if you had some `code here` text and explain in the paragraph below that this is what breaks the stack discipline.

Reply:
After a bit more thought, I'm not sure this will be a problem.  I think we just need to scope the accesses and not try to be too clever with liveness, and stick to a stack discipline then it should be fine.